[1.10] Prevent a crash when scheduling block updates for non-existing blocks

maruohon opened the pull request:
This fixes a crash when trying to load scheduled block updates from Chunk
NBT data (see AnvilChunkLoader#loadEntities()) for blocks that don't exist
anymore for whatever reason (maybe a mod was removed for example).

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/MinecraftForge/MinecraftForge?pullRequest=3485) <br/>All committers have signed the CLA.

mezz commented:
Can you provide a crash report where this happened?

maruohon commented:
User 1): http://pastebin.com/b1PDXhtR
User 2): http://pastebin.com/Us7aLy6V

In User 1's case they removed one mod while updating their map from 1.10.2 to 1.11, and after removing the mod the server would crash every time while trying to load the world.

I actually added a command to one of my mods to selectively remove tile ticks from the world save, and User 1 was able to fix their map with that. In User 2's case it's unknown why the crashes occur, they had updated a mod but not removed one, but maybe the update removed a block...

mezz commented:
Thanks, makes sense.

mezz commented:
This block should really be `Blocks.AIR`, never null. Can you look more into the cause of this issue?

maruohon commented:
Okay so the problem comes from Block.getBlockFromName(). If it is given a string that is not a valid, currently registered block, then it will fail to find it in the map, and thus it falls back to the else case where it tries to parse it as an integer and then get a block by that integer ID. But here is a try-catch, and since the Integer.parseInt() throws an exception, it falls to the catch block where it just returns null.

So should this PR then simply be changed to change that null return into return Blocks.AIR; instead?
Although the Block.getBlockFromName() method is annotated as ```@Nullable```. What is the proper fix then in this case?

![getblockfromname_null](https://cloud.githubusercontent.com/assets/2331050/20862952/946d4946-b9c2-11e6-8123-07263d419456.PNG)

LexManos commented:
The places in which getBlockFromName are used in vanilla all have no null protection. And instead expect AIR for invalid things.
Except for notibly FlatGeneratorInfo, and CommandReplaceItem.
I think for more uniform behavior with the rest of the engine, and to keep patches to a minimal, that return value should change to AIR.
And perhaps modify those two uses to better respect that air is returned, Tho I think CommandReplaceItem is fine.

maruohon committed:
Change Block.getBlockFromName() to return Blocks.AIR on failure

This fixes a NPE crash in WorldServer#scheduleBlockUpdate() when called from
AnvilChunkLoader#loadEntities() on chunks that have saved scheduled block
ticks for block names that no longer exist.

To keep the old behaviour on invalid block names versus a valid air block,
CommandTestForBlock, FlatGeneratorInfo and ItemStack have a couple of
changes where the getBlockFromName() return value null was previosuly used
to indicate an invalid block.

maruohon commented:
I have now changed the PR to return ```Blocks.AIR``` from the ```Block.getBlockFromName()``` method.
I also changed the usage of that returned value in three places, where the old code assumed null for invalid things. The change in ```FlatGeneratorInfo``` looks ugly because I crammed it into two changed lines, but should, and from my testing does maintain the behaviour that Air is still valid for layers, but non-existing strings returning Air fall back to the default flat generator instead. The other thing to note is ```CommandTestForBlock```, where the added String comparisons might affect performance for command-block heavy use cases...

These would really benefit from a separate Blocks.INVALID or something, instead of trying to distinguish between valid and invalid Blocks.AIR...

LexManos commented:
Humm, I need to speak with some people to see what their intetions for this function is.
For the time being i'll take your original quick 'return if null' to fix the crash. 
We'll address the real issue later {because this would crash in many other places as well}

mezz commented:
@LexManos any update on this?
